exercise/static/exercise/chapter.js

```diff
@@ -579,31 +579,100 @@
+        const hashes = new Set();
+        const datahashElements = this.element.find("[data-hash]");
+        for (let elem of datahashElements) {
+          try {
+            hashes.add(elem.dataset["hash"]);
+          } catch(error) {
+          }
         }
-        var url = $(form_element).attr("action");
-        var formData = new FormData(form_element);
 
-        exercise.submitAjax(url, formData, function(data) {
-          //$(form_element).find(":input").prop("disabled", false);
-          //exercise.hideLoader();
-          var input = $(data);
-          if (exercise.quiz) {
-            exercise.quizSubmission("success");
-            exercise.update(input);
-            exercise.renderMath();
-            exercise.dom_element.dispatchEvent(
-              new CustomEvent("aplus:submission-finished",
-                {bubbles: true, detail: {type: exercise.exercise_type}}));
+        var isDuplicateCallback = function(exercise, hash) {
+          isDuplicate = hashes.has(hash);
+          if (isDuplicate) {
+            var continueBtn = document.getElementById("duplicate-submission-modal-button");
+            continueBtn.onclick = () => {
+              submitCallback(exercise, hash);
+            }
+            $("#duplicate-submission-modal").modal()
           } else {
-            exercise.updateSubmission(input);
+            submitCallback(exercise, hash);
           }
-        });
+        }
+
+        // Compute a hash for the current submission
+        const inputFileElements = form_element.querySelectorAll('input[type="file"]');
+
+        if (inputFileElements.length > 0) {
+          const reader = new FileReader();
+          reader.onerror = function() {
+            alert("File cannot be opened");
+          };
+
+          let filesConcatenated = "";
+          let index = 0;
+
+          function readNext(exercise, callback) {
+            const file = inputFileElements[index++].files[0];
+            reader.onload = function() {
+              filesConcatenated += reader.result;
+              if (index < inputFileElements.length) {
+                // More to do, start loading the next file
+                readNext(callback);
+              } else {
+                const hash = md5(filesConcatenated);
+                callback(exercise, hash);
+              }
+            }
+            reader.readAsText(file);
+          };
+
+          readNext(this, isDuplicateCallback);
+        } else if (this.quiz) {
+          const formData = new FormData(form_element);
+          const formProps = Object.fromEntries(formData);
+
+          // Grader language shouldn't affect the hash
+          if (formProps.hasOwnProperty('__grader_lang')) {
+            delete formProps['__grader_lang'];
+          }
+
+          const formDataJsonString = JSON.stringify(formProps, Object.keys(formProps).sort());
+          const hash = md5(formDataJsonString);
+          isDuplicateCallback(this, hash);
```

Comment:
Most of the new code here in chapter.js should be moved to a new JS file. Define functions and call the functions here in chapter.js. It will make the code cleaner and chapter.js won't look so bloated. The duplicate check functions would also be more reusable and they could be used outside the chapter.js code too.